File: go/who_took_test.go
package main

import (
	. "github.com/onsi/ginkgo"
	. "github.com/onsi/gomega"
)

var null = Null
var hearts = Hearts
var spades = Spades

var _ = Describe("Goker", func() {
	It("should win Ace (in same colors)", func() {
		cards := []DownCard{
			{Hearts, _7, nil},
			{Hearts, Ace, nil},
			{Hearts, King, nil},
			{Hearts, Jack, nil},
		}

		Expect(whoTook(cards, Hearts, 0)).To(Equal(1))
		Expect(whoTook(cards, Hearts, 1)).To(Equal(1))
		Expect(whoTook(cards, Hearts, 2)).To(Equal(1))
		Expect(whoTook(cards, Hearts, 3)).To(Equal(1))
		Expect(whoTook(cards, Clubs, 3)).To(Equal(1))
		Expect(whoTook(cards, Diamonds, 3)).To(Equal(1))
		Expect(whoTook(cards, Spades, 3)).To(Equal(1))
		Expect(whoTook(cards, None, 3)).To(Equal(1))
	})

	It("should win trump", func() {
		cards := []DownCard{
			{Hearts, _7, nil},
			{Hearts, Ace, nil},
			{Diamonds, _6, nil},
			{Hearts, Jack, nil},
		}

		Expect(whoTook(cards, Diamonds, 0)).To(Equal(2))
		Expect(whoTook(cards, Diamonds, 1)).To(Equal(2))
		Expect(whoTook(cards, Diamonds, 2)).To(Equal(2))
		Expect(whoTook(cards, Diamonds, 3)).To(Equal(2))
	})

	It("should win first card (in no color)", func() {
		cards := []DownCard{
			{Clubs, _7, nil},
			{Diamonds, Ace, nil},
			{Hearts, _6, nil},
			{Spades, Jack, nil},
		}

		Expect(whoTook(cards, None, 0)).To(Equal(0))
		Expect(whoTook(cards, None, 1)).To(Equal(1))
		Expect(whoTook(cards, None, 2)).To(Equal(2))
		Expect(whoTook(cards, None, 3)).To(Equal(3))
	})

	It("should handle special case", func() {
		cards := []DownCard{
			{Hearts, _6, nil},
			{Hearts, _8, nil},
			{Hearts, _7, nil},
			{Hearts, _10, nil},
		}

		Expect(whoTook(cards, None, 2)).To(Equal(3))
	})

	It("should handle special case 2", func() {
		cards := []DownCard{
			{Diamonds, _6, nil},
			{Clubs, _7, nil},
			{Clubs, _9, nil},
			{Diamonds, Queen, nil},
		}

		Expect(whoTook(cards, Clubs, 2)).To(Equal(2))
	})

	It("should let joker win", func() {
		cards := []DownCard{
			{Diamonds, Queen, nil},
			{Spades, Queen, nil},
			{Hearts, _6, nil},
			{Spades, King, nil},
		}

		Expect(whoTook(cards, None, 0)).To(Equal(0))
	})

	It("should let second joker win", func() {
		cards := []DownCard{
			{Clubs, _6, &JokerSubAction{true, &null}},
			{Hearts, _7, nil},
			{Spades, _6, &JokerSubAction{true, &null}},
			{Diamonds, _6, nil},
		}

		Expect(whoTook(cards, Clubs, 3)).To(Equal(2))
	})

	It("should give somebody else if joker does not want it", func() {
		cards := []DownCard{
			{Diamonds, _10, nil},
			{Clubs, _6, &JokerSubAction{false, &hearts}},
			{Hearts, King, nil},
			{Spades, _8, nil},
		}

		Expect(whoTook(cards, None, 1)).To(Equal(2))
	})

	It("should give somebody else if neither want it", func() {
		cards := []DownCard{
			{Clubs, _9, nil},
			{Clubs, _6, &JokerSubAction{false, &hearts}},
			{Spades, _6, &JokerSubAction{false, &hearts}},
			{Hearts, _6, nil},
		}

		Expect(whoTook(cards, None, 2)).To(Equal(3))
	})

	It("should handle case when joker does not want trump", func() {
		cards := []DownCard{
			{Diamonds, _10, nil},
			{Spades, _6, &JokerSubAction{true, &hearts}},
			{Hearts, King, nil},
			{Spades, _8, nil},
		}

		Expect(whoTook(cards, Diamonds, 1)).To(Equal(0))
		Expect(whoTook(cards, None, 1)).To(Equal(1))
		Expect(whoTook(cards, Clubs, 1)).To(Equal(1))
	})

	It("should let joker take it even if he does not want it", func() {
		cards := []DownCard{
			{Diamonds, _10, nil},
			{Spades, _6, &JokerSubAction{false, &hearts}},
			{Spades, King, nil},
			{Spades, _8, nil},
		}

		Expect(whoTook(cards, None, 1)).To(Equal(1))
		Expect(whoTook(cards, Clubs, 1)).To(Equal(1))
		Expect(whoTook(cards, Diamonds, 1)).To(Equal(0))
	})

	It("should let somebody else take if if joker don't want it", func() {
		cards := []DownCard{
			{Diamonds, _10, nil},
			{Spades, _6, &JokerSubAction{false, &hearts}},
			{Spades, King, nil},
			{Spades, _8, nil},
		}

		Expect(whoTook(cards, Diamonds, 1)).To(Equal(0))
		Expect(whoTook(cards, Spades, 1)).To(Equal(2))
	})

	It("should handle when joker does not want and trump card takes it", func() {
		cards := []DownCard{

			{Diamonds, _10, nil},
			{Spades, _6, &JokerSubAction{false, &spades}},
			{Clubs, King, nil},
			{Spades, _8, nil},
		}

		Expect(whoTook(cards, Clubs, 1)).To(Equal(2))
	})
})

File: node-c/test/whotook.test.ts
import { CardColor, CardLevel, DownCard } from '../lib/types';
import { getCDownCardFromDownCard } from '../lib/type-utils';
import { whoTook1, whoTook2 } from '../lib/jok';

function whoTook2Wrapper(
  cards: DownCard[],
  trumpColor: CardColor,
  firstCardIndex: number,
) {
  const cDownCards = cards.map(getCDownCardFromDownCard);

  return whoTook2(
    cDownCards[0],
    cDownCards[1],
    cDownCards[2],
    cDownCards[3],
    trumpColor,
    firstCardIndex);
}

const testCases = {
  whoTook1: whoTook1,
  whoTook2: whoTook2Wrapper
};

for (let entry of Object.entries(testCases)) {
  describe(entry[0], () => {
    const whoTook = entry[1];

    it('should win Ace (in same colors)', () => {
      const cards: DownCard[] = [
        [CardColor.Hearts, CardLevel._7],
        [CardColor.Hearts, CardLevel.Ace],
        [CardColor.Hearts, CardLevel.King],
        [CardColor.Hearts, CardLevel.Jack],
      ];

      expect(whoTook(cards, CardColor.Hearts, 0)).toBe(1);
      expect(whoTook(cards, CardColor.Hearts, 1)).toBe(1);
      expect(whoTook(cards, CardColor.Hearts, 2)).toBe(1);
      expect(whoTook(cards, CardColor.Hearts, 3)).toBe(1);
      expect(whoTook(cards, CardColor.Clubs, 3)).toBe(1);
      expect(whoTook(cards, CardColor.Diamonds, 3)).toBe(1);
      expect(whoTook(cards, CardColor.Spades, 3)).toBe(1);
      expect(whoTook(cards, CardColor.None, 3)).toBe(1);
    });

    it('should win trump', () => {
      const cards: DownCard[] = [
        [CardColor.Hearts, CardLevel._7],
        [CardColor.Hearts, CardLevel.Ace],
        [CardColor.Diamonds, CardLevel._6],
        [CardColor.Hearts, CardLevel.Jack],
      ];

      expect(whoTook(cards, CardColor.Diamonds, 0)).toBe(2);
      expect(whoTook(cards, CardColor.Diamonds, 1)).toBe(2);
      expect(whoTook(cards, CardColor.Diamonds, 2)).toBe(2);
      expect(whoTook(cards, CardColor.Diamonds, 3)).toBe(2);
    });

    it('should win first card (in no color)', () => {
      const cards: DownCard[] = [
        [CardColor.Clubs, CardLevel._7],
        [CardColor.Diamonds, CardLevel.Ace],
        [CardColor.Hearts, CardLevel._6],
        [CardColor.Spades, CardLevel.Jack],
      ];

      // always trump is the winner
      expect(whoTook(cards, CardColor.None, 0)).toBe(0);
      expect(whoTook(cards, CardColor.None, 1)).toBe(1);
      expect(whoTook(cards, CardColor.None, 2)).toBe(2);
      expect(whoTook(cards, CardColor.None, 3)).toBe(3);
    });

    it('special case', () => {
      const cards: DownCard[] = [
        [CardColor.Hearts, CardLevel._6],
        [CardColor.Hearts, CardLevel._8],
        [CardColor.Hearts, CardLevel._7],
        [CardColor.Hearts, CardLevel._10],
      ];

      // always trump is the winner
      expect(whoTook(cards, CardColor.None, 2)).toBe(3);
    });

    it('special case2', () => {
      const cards: DownCard[] = [
        [CardColor.Diamonds, CardLevel._6],
        [CardColor.Clubs, CardLevel._7],
        [CardColor.Clubs, CardLevel._9],
        [CardColor.Diamonds, CardLevel.Queen],
      ];

      // always trump is the winner
      expect(whoTook(cards, CardColor.Clubs, 2)).toBe(2);
    });

    it('When Trump is Joker', () => {
      const cards: DownCard[] = [
        [CardColor.Diamonds, CardLevel.Queen],
        [CardColor.Spades, CardLevel.Queen],
        [CardColor.Hearts, CardLevel._6],
        [CardColor.Spades, CardLevel.King],
      ];

      // always trump is the winner
      expect(whoTook(cards, CardColor.None, 0)).toBe(0);
    });

    it('Double Jokers wants', () => {
      const cards: DownCard[] = [
        [
          CardColor.Clubs,
          CardLevel._6,
          { want: true, color: CardColor.Null },
        ],
        [CardColor.Hearts, CardLevel._7],
        [
          CardColor.Spades,
          CardLevel._6,
          { want: true, color: CardColor.Null },
        ],
        [CardColor.Diamonds, CardLevel._7],
      ];

      // always trump is the winner
      expect(whoTook(cards, CardColor.Clubs, 3)).toBe(2);
    });

    it('Joker dont want', () => {
      const cards: DownCard[] = [
        [CardColor.Diamonds, CardLevel._10],
        [
          CardColor.Clubs,
          CardLevel._6,
          { want: false, color: CardColor.Hearts },
        ],
        [CardColor.Hearts, CardLevel.King],
        [CardColor.Spades, CardLevel._8],
      ];

      // always trump is the winner
      expect(whoTook(cards, CardColor.None, 1)).toBe(2);
    });

    it('Double Jokers dont want', () => {
      const cards: DownCard[] = [
        [CardColor.Clubs, CardLevel._9],
        [CardColor.Clubs, CardLevel._6, { want: false, color: -1 }],
        [
          CardColor.Spades,
          CardLevel._6,
          { want: false, color: CardColor.Hearts },
        ],
        [CardColor.Hearts, CardLevel._6],
      ];

      // always trump is the winner
      expect(whoTook(cards, CardColor.None, 2)).toBe(3);
    });

    it('Joker wants non-Trump', () => {
      const cards: DownCard[] = [
        [CardColor.Diamonds, CardLevel._10],
        [
          CardColor.Spades,
          CardLevel._6,
          { want: true, color: CardColor.Hearts },
        ],
        [CardColor.Hearts, CardLevel.King],
        [CardColor.Spades, CardLevel._8],
      ];

      expect(whoTook(cards, CardColor.Diamonds, 1)).toBe(0);
      expect(whoTook(cards, CardColor.None, 1)).toBe(1);
      expect(whoTook(cards, CardColor.Clubs, 1)).toBe(1);
    });

    it('Joker wants non-Trump, but another Joker wants', () => {
      const cards: DownCard[] = [
        [
          CardColor.Spades,
          CardLevel._6,
          { want: true, color: CardColor.Diamonds },
        ],
        [CardColor.Clubs, CardLevel._10],
        [CardColor.Clubs, CardLevel._6, { want: true, color: null }],
        [CardColor.Diamonds, CardLevel._8],
      ];

      expect(whoTook(cards, CardColor.Clubs, 0)).toBe(2);
    });

    it("Joker don't want but still takes", () => {
      const cards: DownCard[] = [
        [CardColor.Diamonds, CardLevel._10],
        [
          CardColor.Spades,
          CardLevel._6,
          { want: false, color: CardColor.Hearts },
        ],
        [CardColor.Spades, CardLevel.King],
        [CardColor.Spades, CardLevel._8],
      ];

      expect(whoTook(cards, CardColor.None, 1)).toBe(1);
      expect(whoTook(cards, CardColor.Clubs, 1)).toBe(1);
      expect(whoTook(cards, CardColor.Diamonds, 1)).toBe(0);
    });

    it("Joker don't want", () => {
      const cards: DownCard[] = [
        [CardColor.Diamonds, CardLevel._10],
        [
          CardColor.Spades,
          CardLevel._6,
          { want: false, color: CardColor.Hearts },
        ],
        [CardColor.Spades, CardLevel.King],
        [CardColor.Spades, CardLevel._8],
      ];

      expect(whoTook(cards, CardColor.Diamonds, 1)).toBe(0);
      expect(whoTook(cards, CardColor.Spades, 1)).toBe(2);
    });

    it("Joker don't want and takes Trump", () => {
      const cards: DownCard[] = [
        [CardColor.Diamonds, CardLevel._10],
        [
          CardColor.Spades,
          CardLevel._6,
          { want: false, color: CardColor.Spades },
        ],
        [CardColor.Clubs, CardLevel.King],
        [CardColor.Spades, CardLevel._8],
      ];

      expect(whoTook(cards, CardColor.Clubs, 1)).toBe(2);
    });
  });
}
